### ISSUE TITLE
tuple-like key_value_pair (for structured bindings)

### ISSUE THREAD

**Issue by AeroStun** [CPPAlliance/json#111]
At the moment `key_value_pair` cannot be used for structured bindings without user boilerplate.
Since it does model a pair anyway and offers an interface similar to `std::pair`, it would be nice for it to also specialize `std::tuple_size`, `std::tuple_element` and a form of `get` when using C++17 or above
This would make the following code well-formed in C++17 or above:
```cpp
object o{{"one", 1}, {"two", 2}};
const auto it = o.find("one");
if (it == o.end())
  handle_error();
const auto& [key, value] = *it;
```
Note that replacing `object` with `std::map<std::string, int>` would produce well-formed C++17 code (see https://godbolt.org/z/fVZUkC)

If this is something the authors would like too see too, I'd happily make a PR for it

**Comment by vinniefalco** [comment]
This needs:
1. To be conditionally enabled when C++17 or later
2. To have tests, also conditionally enabled

@AeroStun up to you if you want to try it as a pull request, but if not then @sdkrystian will do it.

**Comment by AeroStun** [comment]
I'll give it a try right now; might come back with questions if I fail to find an answer by reading the current code

**Comment by sdkrystian** [comment]
We don't need this conditionally enabled.

**Comment by vinniefalco** [comment]
Ah <tuple>is C++11. Hmm... including <tuple> from value.hpp is heavyweight. Maybe this should go in a different header so that someone can avoid getting those headers if they don't need it.

**Comment by AeroStun** [comment]
Well, given `<array>` and `<utility>` contain specializations of `std::tuple_element` and `std::tuple_size`, isn't it fair to assume that we can rely on those being included instead? (they are already included in `value.hpp`)

**Comment by vinniefalco** [comment]
If true, that's a good point! I don't recall why I include `<array>` though, doesn't sound needed.

**Comment by AeroStun** [comment]
Self-correction: you don't include `<array>`, only `<utility>`

**Comment by vinniefalco** [comment]
Yes you can specialize it after including `<utility>` good idea!

**Comment by sdkrystian** [comment]
I have an implementation done  -- not sure if we want to document all this boilerplate.

**Comment by vinniefalco** [comment]
We don't need to include std symbols in the reference, but we should include `json::get` in the reference, and in the exposition we should point out that structured bindings will work. Perhaps in a sub-section with the heading "C++17 and later."

**Comment by AeroStun** [comment]
I have <https://github.com/AeroStun/json/commit/5f4d3fca4e4fa5112bdcbf928a5b3c15872aa751> running some CIs

**Comment by vinniefalco** [comment]
@sdkrystian when a user offers to submit a contribution, let them :)

**Comment by vinniefalco** [comment]
@AeroStun I left comments: https://github.com/AeroStun/json/commit/5f4d3fca4e4fa5112bdcbf928a5b3c15872aa751

**Comment by AeroStun** [comment]
@vinniefalco Addressed comments: https://github.com/AeroStun/json/commit/333e3bb5b885e5ca97d3c8fe49155f191a1eca6d

**Comment by AeroStun** [comment]
I'm pretty much ready to squash and PR except clang-3.8 on Travis fails, likely due to a compiler bug
https://travis-ci.com/github/AeroStun/json/jobs/348807479#L1276
Googling around seems to suggest that this is related to https://bugs.llvm.org/show_bug.cgi?id=17537#c4
How do you want to handle old compilers having nasty bugs?

**Comment by vinniefalco** [comment]
Well it depends. Does Boost.JSON support clang-3.8?

**Comment by vinniefalco** [comment]
Unfortunately it does (according to the docs)

>Tested with these compilers: msvc-14.1+, gcc 4.8.4+, clang 3.6+.

We have to figure out a work-around.

**Comment by vinniefalco** [comment]
Please submit the branch as a pull request

**Comment by AeroStun** [comment]
I found a workaround in the meantime
Going to squash and PR

**Comment by sdkrystian** [comment]
Addressed by #112
